numpy/f2py/cfuncs.py

```diff
@@ -488,9 +488,9 @@
         char *_from = (from);                                   \\
         FAILNULL(_to); FAILNULL(_from);                         \\
         (void)strncpy(_to, _from, sizeof(char)*_m);             \\
-        _to[_m-1] = '\\0';                                      \\
+        _to[_m] = '\\0';                                        \\
```

Comment:
This change looks wrong to me, the call on line 627 results in a write to `PyArray_DATA(arr)[PyArray_NBYTES(arr)]` which is beyond the end of the array.

Reply:
Perhaps line 627 is wrong instead.

Reply:
Hmm, it is possible that the small array cache is hiding an invalid write, or the tests don't cover 627? (if that is the case, need to set these to 0, but I can try that):
https://github.com/numpy/numpy/blob/55ffea92285dfe5b49123c57b63dcd6f00c5f3c3/numpy/core/src/multiarray/alloc.c#L34-L36)

Curious, the below replaces NULL termination with space, is that for copying in both directions (mainly curious)?  Other than that, line avoiding the NULL padding when copying into an array is the only concern.  I wouldn't be surprised if just using a plain `strncpy` is correct there.

Reply:
I think line 627 is never even emitted by f2py; the entire function that line is in seems suspect. Can we just remove it entirely?

Reply:
> I think line 627 is never even emitted by f2py; the entire function that line is in seems suspect. Can we just remove it entirely?

Not true, it is emitted from https://github.com/numpy/numpy/blob/0fe69ae3fc513aefd32a58ae8ccac294b12300d2/numpy/f2py/rules.py#L966

Reply:
Then it sounds like line 627 needs fixing, perhaps by calling `strncopy` directly.

Reply:
Yes, I'll cook up an example that touches this code and then look for a fix.

Reply:
Here's a MWE:

```python
code = """
subroutine test_inout(a)
  implicit none
  character(len=4), intent(inout) :: a
  a(1:1) = 'A'
end subroutine test_inout
"""

from numpy import array
import numpy.f2py

numpy.f2py.compile(code, modulename='sizeinout', extension='.f90', verbose=0, extra_args="--debug-capi")

import sizeinout

a = array(b'1234')
sizeinout.test_inout(a)
print(f"a = {a}")
```